File: 0918-maximum-sum-circular-subarray/0918-maximum-sum-circular-subarray.c
int maxSubarraySumCircular(int* nums, int numsSize){
    int curr_max, max_val, curr_min, min_val, sum;
    curr_max = max_val = -40000;
    curr_min = min_val = 40000;
    sum = 0;
    for(int i = 0; i < numsSize; i++){
        curr_max = nums[i] > nums[i] + curr_max ? nums[i] : nums[i] + curr_max;
        curr_min = nums[i] > nums[i] + curr_min ? nums[i] + curr_min : nums[i];
        max_val = max_val > curr_max ? max_val : curr_max;
        min_val = min_val < curr_min ? min_val : curr_min;
        sum += nums[i];
    }
    if(sum == min_val){
        return max_val;
    }
    return max_val > sum - min_val ? max_val : sum - min_val;
}

File: 0918-maximum-sum-circular-subarray/0918-maximum-sum-circular-subarray.py
class Solution:
    def maxSubarraySumCircular(self, nums: List[int]) -> int:
        maxSub, minSub = [float("-inf")] * len(nums), [float("inf")] * len(nums)
        maxSub[0] = minSub[0] = nums[0]
        for i in range(1, len(nums)):
            maxSub[i], minSub[i] = max(nums[i], nums[i] + maxSub[i - 1]), min(nums[i], nums[i] + minSub[i - 1])
        maxVal, minVal = max(maxSub), min(minSub)
        if sum(nums) - minVal == 0:
            # entire array is negative, return max of subarry
            return maxVal
        return max(maxVal, sum(nums) - minVal)
            